docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,329 @@
+
+The central data structure representing Celestia block data is EDS(`rsmt2d.ExtendedDataSquare`) and the new storage design
+is focused around storing entire EDSes as a whole rather than a set of individual chunks s.t. storage subsystem
+can handle storing and streaming/serving blocks of 4mb sizes and more.
+
+##### EDS Serde
+
+Storing EDS as a whole requires EDS (de)serialization. For this the [CAR format](https://ipld.io/specs/transport/car) is
+chosen.
+
+###### `share.WriteEDS`
+
+To write EDS into a stream/file `WriteEDS` is introduced.
+// TODO Describe logic
+
+NOTE: CAR provides [a utility](https://github.com/ipld/go-car/blob/master/car.go#L47) to serialize any DAG into the file and
+there is a way to serialize EDS into DAG(`share/ipld.ImportShares`). This approach is the simplest and traverses shares
+and Merkle Proofs in depth-first manner packing them in a CAR file. However, this is incompatible with the requirement
+to truncate the CAR file to get strictly the first quadrant out of it without NMT proofs, so serialization must be
+different from the utility to support that.
+
+NOTE2: Alternatively to `WriteEDS`, and `EDSReader` could be introduced to make EDS-to-stream handling more idiomatic
+and efficient in some cases, with the cost of more complex implementation.
+
+```go
+// WriteEDS writes whole EDS into given io.Writer as CARv1 file.
+// All its shares and recomputed NMT proofs.
+func WriteEDS(context.Context, *rsmt2d.ExtendedDataSquare, io.Writer) error
+```
+
+###### `share.ReadEDS`
+
+To read EDS out of stream/file `ReadEDS` is introduced. Internally, it
+
+- Imports EDS with an empty pre-allocated slice. NOTE: Size can be taken from DataRoot
+- Wraps given io.Reader with [`BlockReader`](https://github.com/ipld/go-car/blob/master/v2/block_reader.go#L16)
+- Reads out blocks one by one and fills up the EDS quadrant via `EDS.SetCell`
+- Recomputes and validates via `EDS.Repair`
+
+```go
+// ReadEDS reads EDS quadrant(1/4) from io.Reader CAR file.
+//
+// It expects strictly first EDS quadrant(top left).
+// Returned EDS is guaranteed to be full and valid against DataRoot, otherwise ReadEDS errors.
+func ReadEDS(context.Context, io.Reader, DataRoot) (*rsmt2d.ExtendedDataSquare, error)
+```
+
+##### `share.EDSStore`
+
+To manage every EDS on the disk FNs/BNs keep `EDSStore` type is introduced in `share` pkg. Each EDS together with its
+Merkle Proofs serializes into CARv1 file. All the serialized CARv1 file blobs are stored as OS FS files as DAGStore
+[Mounts](https://github.com/filecoin-project/dagstore/blob/master/mount/mount.go).
+
+The introduced `EDSStore` also maintains (via DAGStore) a top-level index enabling granular and efficient random access
+to every share and/or Merkle proof over every registered CARv1 file. The `EDSStore` provides a custom `Blockstore` interface
+implementation to achieve the access. The main use-case is randomized sampling over the whole chain of EDS block data
```

Comment:
We should mention that the Blockstore interface implementation is only partial

Reply:
If we use the one from DAGStore, then it is not.

Reply:
It is tho 

Not all methods are supported operations (put doesn't make any sense for example)

Reply:
> > implementation is only partial
> If we use the one from DAGStore, then it is not.
This is about being partial or not. DAGStore implements everything. The fact that we don't need Put from it seems orthogonal. We _could_ mention, though that we don't need it there, I agree

Reply:
My reasoning for this was just that if we used their implementation (at least in my experiments), it wouldn't work, because they panic on Put, which we never call explicitly but the exchange on the BlockService calls it on Add

But this is besides the point here and I think we can resolve this discussion

Reply:
I will double-check the necessity for the Put. Good point

Reply:
@distractedm1nd. The Put is not suitable for our Blockstore, because of the WriteEDS specifics. We could make Retriever(in the fallback case) write into the specific CAR file and its blockstore, but that breaks the quadrant ordering of writes, meaning that `ReadEDS` would read out a corrupted quadrant. Hence we should stick to the DAGStore' Blockstore, unless the caching issues you've mentioned is harmful to us